In [1]:
!pip install numpy


[notice] A new release of pip available: 22.3 -> 22.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import numpy

In [3]:
numpy.__version__

'1.23.4'

In [4]:
import numpy as np

# Упражнения по библиотеке Numpy

# 1. Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [5]:
a = np.arange(11)
a[(3 < a) & (a < 8)] *= -1
print(a)

[ 0  1  2  3 -4 -5 -6 -7  8  9 10]


In [6]:
a = np.arange(10)
np.where((a > 3) & (a < 8), a*(-1), a)

array([ 0,  1,  2,  3, -4, -5, -6, -7,  8,  9])

# 2. Заменить максимальный элемент случайного массива на 0

In [7]:
a = np.random.random(15)


In [8]:
print(a)

[0.91771219 0.73379488 0.74131511 0.28698895 0.15053562 0.18478955
 0.9181364  0.59738173 0.32626029 0.10355756 0.53075915 0.81838976
 0.48289747 0.03759367 0.71059752]


In [9]:
a[a.argmax()] = 0
print(a)

[0.91771219 0.73379488 0.74131511 0.28698895 0.15053562 0.18478955
 0.         0.59738173 0.32626029 0.10355756 0.53075915 0.81838976
 0.48289747 0.03759367 0.71059752]


# 3. Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [10]:
def cartesian(arrays):
    arrays = [np.asarray(a) for a in arrays]
    shape = (len(x) for x in arrays)

    ix = np.indices(shape, dtype=int)
    ix = ix.reshape(len(arrays), -1).T

    for n, arr in enumerate(arrays):
        ix[:, n] = arrays[n][ix[:, n]]

    return ix

print (cartesian(([1, 2, 3], [4, 5], [6, 7])))

[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]


# 4. Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [11]:
a = np.random.randint(0,5,(8,3))
b = np.random.randint(0,5,(2,2))

d = (a[..., np.newaxis, np.newaxis] == b)
rows = np.where(d.any((3,1)).all(1))[0]
print(rows)

[0 1 3 5 6 7]


# 5. Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [12]:
a = np.random.randint(0,5,(10,3))
print(a)

b = np.all(a[:,1:] == a[:,:-1], axis=1)
c = a[~b]
print(c)

c = a[a.max(axis=1) != a.min(axis=1),:]
print(c)


[[1 1 3]
 [3 2 1]
 [3 3 4]
 [2 1 3]
 [3 1 4]
 [0 1 4]
 [1 3 2]
 [4 3 0]
 [2 0 0]
 [0 1 1]]
[[1 1 3]
 [3 2 1]
 [3 3 4]
 [2 1 3]
 [3 1 4]
 [0 1 4]
 [1 3 2]
 [4 3 0]
 [2 0 0]
 [0 1 1]]
[[1 1 3]
 [3 2 1]
 [3 3 4]
 [2 1 3]
 [3 1 4]
 [0 1 4]
 [1 3 2]
 [4 3 0]
 [2 0 0]
 [0 1 1]]


# 6. Дан двумерный массив. Удалить те строки, которые повторяются

In [13]:
a = np.array([[1, 1], [2, 3]])
np.unique(a)

array([1, 2, 3])

# Задача 1: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.
Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [14]:
x = np.array([[1,0,1],[2,0,2],[3,0,3],[4,4,4]])
print(x)

[[1 0 1]
 [2 0 2]
 [3 0 3]
 [4 4 4]]


In [15]:

np.prod(np.diag(x)[np.diag(x)!=0]  )

3

# Задача 2: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [15]:
import numpy as np

In [16]:
x = np.array([1,2,2,4])
y = np.array([4,2,1,2])

mult = np.isin(x, y)

In [17]:
mult

array([ True,  True,  True,  True])

2 cпособ

In [10]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
x.sort() == y.sort()

True

# Задача 3: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль.
Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [18]:
x = np.array([6,2,0,3,0,0,5,7,0])
null = x == 0
print (x[1:][null[:-1]].max())

5


2 способ

In [11]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max([x[i] for i in range(1, len(x)) if x[i - 1] == 0])

5

# Задача 4: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.
Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [12]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1], dtype=int64))

2 способ 

In [13]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
result = {}
for element in x:
    result[element] = result.get(element, 0) + 1
values = [key for key in result.keys()]
val_frequency = [values for values in result.values()]
result_coding = (values, val_frequency)
result_coding

([2, 3, 5], [3, 3, 1])

# Задача 5: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [21]:
import numpy as np

p1 = np.array((1, 2, 3))
p2 = np.array((4, 5, 6))

In [22]:
import math
distance = math.dist(p1, p2)
print(distance)

5.196152422706632


In [23]:
import numpy as np

p1 = np.array((1, 2, 3))
p2 = np.array((4, 5, 6))

In [24]:
distance = np.sqrt(np.dot(p1-p2, p1-p2))
print(distance)

5.196152422706632


# Задача 6: CrunchieMunchies *
Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием CrunchieMunchies.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши СrunchieMunchies - самый здоровый выбор для потребителей.

In [12]:
import numpy as np

Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [16]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

 2.В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?
Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [17]:
average_calories = np.average(calorie_stats) - 60
average_calories

46.883116883116884

 3.Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.
Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [18]:
calorie_stats_sorted = np.sort(calorie_stats)
len(calorie_stats_sorted)

77

 4.Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.
Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [19]:
median_calories = np.median(calorie_stats)
median_calories

110.0

 5.В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.
Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [20]:
nth_percentile = 0
while np.percentile(calorie_stats, nth_percentile) <= 60:
    nth_percentile += 1
nth_percentile

4

 6.Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.
Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [21]:
more_calories = len(calorie_stats[calorie_stats > 60]) / len(calorie_stats)
more_calories

0.961038961038961

 7.Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?
Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [22]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

 8.Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Thus, the following conclusions can be drawn:
The sample consists of 77 indicators of the calorie content of cereals of different brands, min = 50, max = 160.
The average calorie count for CrunchieMunchies is higher at 46.
CrunchieMunchies flakes are lower in calories than 96% of competitors' flakes.